### ISSUE TITLE
bug in path type coercion?

### ISSUE THREAD

**Issue by ecavallo** [RedPRL/redtt#228]
```
import path

let thing (A : type) (a,b : A) (p : Path A a b) : Path (Path A a b) p p =
  λ i j →
  let ty : dim → type = λ k →
    let h : dim → A = λ l →
      comp 0 l a [
      | k=0 ⇒ λ z → p z
      | k=1 ⇒ λ z → p z
      ]
    in
    Path (Path A (h 0) (h 1)) (λ l → h l) (λ l → h l)
  in
  (coe 0 1 (λ _ z → p z) in ty) i j
```
raises
```
rigid-rigid mismatch: (hcom 0 j
                       A
                       a
                       [%6539=0 <x0> (@ p x0)] [%6539=1 <x0> (@ p x0)]) ≐
                       (@ p j) : A
Encountered error:
    Failure("rigid-rigid mismatch")
```
No clue what's going wrong, but I _think_ this should typecheck...

**Comment by ecavallo** [comment]
Been trying, but it's resilient...

**Comment by jonsterling** [comment]
@ecavallo I played with it a bit and it appears to be some stupid issue with the elaborator rather than the core. That is, if you wrap the last line in `?{ }`, then it works; what `?{ ... }` does is apply the non-invertible introduction rule for a restriction pretype. What it does is tries to produce an element of the carrier type, and *then* checks the boundary.

This should really be equivalent, but I guess the elaborator is dumb.

**Comment by jonsterling** [comment]
Oh shoot, ignore my last comment (deleting).

**Comment by jonsterling** [comment]
Could you explain why this should be well-typed? It's too complicated for me to be confident in reasoning through it myself, and I stepped through the goals a bit, but they were also too long to actually read.

**Comment by ecavallo** [comment]
Here's something which is slightly simpler (one less layer of paths, and separated in two steps):
```
import path

let thing0 (A : type) (B : (dim → A) → type) (p : dim → A) (b : B p) : B p =
  let ty : dim → type = λ k →
    let h : dim → A = λ l →
      comp 0 l (p 0) [
      | k=0 ⇒ λ z → p z
      | k=1 ⇒ λ z → p z
      ]
    in
    B (λ l → h l) 
  in
  (coe 0 1 b in ty)

let thing1 (A : type) (a : A) : Path A a a =
  λ i →
  thing0 A
    (λ q → Path A (q 0) (q 1))
    (λ _ → a)
    (λ _ → a)
    i
```
The first term typechecks, but fails when we instantiate it with a line of path types.

**Comment by ecavallo** [comment]
Note: the error disappears if i contract the `lam i -> ... i` in the second definition.

**Comment by jonsterling** [comment]
Fixed in #230 (i.e. these examples now work; the general fix is proposed in #231, which #230 is trying to work towards)
